testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java

```diff
@@ -587,6 +589,63 @@ public void clientCancel() throws Exception {
     verify(mockClientStreamListener, never()).closed(any(Status.class), any(Metadata.class));
   }
 
+  /**
+   * Test for https://github.com/grpc/grpc-java/issues/1682
+   */
+  @Test(timeout = 5000)
+  public void clientCancelFromWithinMessageRead() throws Exception {
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    serverTransport = serverTransportListener.transport;
+
+    final AtomicBoolean closedCalled = new AtomicBoolean();
+    final ClientStream clientStream = client.newStream(methodDescriptor, new Metadata());
+    clientStream.start(new ClientStreamListener() {
+      @Override
+      public void headersRead(Metadata headers) {
+      }
+
+      @Override
+      public void closed(Status status, Metadata trailers) {
+        assertEquals(Status.CANCELLED.getCode(), status.getCode());
+        assertEquals("nevermind", status.getDescription());
+        closedCalled.set(true);
+      }
+
+      @Override
+      public void messageRead(InputStream message) {
+        assertEquals("foo", methodDescriptor.parseResponse(message));
+        clientStream.cancel(Status.CANCELLED.withDescription("nevermind"));
+      }
+
+      @Override
+      public void onReady() {
+      }
+    });
+    clientStream.halfClose();
+    clientStream.request(1);
+
+    StreamCreation serverStreamCreation
+        = serverTransportListener.takeStreamOrFail(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    assertEquals(methodDescriptor.getFullMethodName(), serverStreamCreation.method);
+    ServerStream serverStream = serverStreamCreation.stream;
+    ServerStreamListener mockServerStreamListener = serverStreamCreation.listener;
+    verify(mockServerStreamListener, timeout(TIMEOUT_MS)).onReady();
+
+    assertTrue(serverStream.isReady());
+    serverStream.writeHeaders(new Metadata());
+    serverStream.writeMessage(methodDescriptor.streamRequest("foo"));
+    serverStream.flush();
+
+    while (!closedCalled.get()) {
+      LockSupport.parkNanos(TimeUnit.MILLISECONDS.toNanos(100));
```

Comment:
LockSupport is really intended for building locking constructs.  This code would be simpler if you just used SettableFuture above, and hung on get() here.